File: oma-ts/src/core/layaClassifier.ts
/**
 * OMA Laya Task Classifier -- Local System 1 Decision & Task Encapsulation.
 * TypeScript implementation adhering to OMA Invariant 1 (serializable state).
 */

export enum TaskCategory {
  CODING = 'coding',
  RESEARCH = 'research',
  MATH_LOGIC = 'math_logic',
  AUTOMATION = 'automation',
  GENERAL = 'general',
}

export enum TaskComplexity {
  SIMPLE = 'simple',
  MODERATE = 'moderate',
  COMPLEX = 'complex',
}

export enum RoutingDecision {
  EXECUTE_DIRECT = 'execute_direct',
  DECOMPOSE_SUBAGENTS = 'decompose_subagents',
  RALPH_LOOP = 'ralph_loop',
  TIERED_FAILOVER = 'tiered_failover',
}

export interface SubTaskSpec {
  id: string;
  role: string;
  sub_agent_type: string;
  objective: string;
  expected_output: string;
  priority: number;
}

export interface TaskEncapsulationData {
  task_id: string;
  objective: string;
  category: TaskCategory;
  complexity: TaskComplexity;
  assigned_agent: string;
  sub_agents: SubTaskSpec[];
  routing_decision: RoutingDecision;
  confidence: number;
  recommended_provider: string;
  criteria: Record<string, unknown>;
  context: string;
  metadata: Record<string, unknown>;
  created_at: number;
}

export interface GateResult {
  passed: boolean;
  confidence: number;
  decision: 'PASSED' | 'FAILOVER';
  score: number;
  notes: string;
  evaluator: string;
}

export class TaskEncapsulation {
  constructor(public data: TaskEncapsulationData) {}

  get taskId(): string { return this.data.task_id; }
  get objective(): string { return this.data.objective; }
  get category(): TaskCategory { return this.data.category; }
  get complexity(): TaskComplexity { return this.data.complexity; }
  get assignedAgent(): string { return this.data.assigned_agent; }
  get subAgents(): SubTaskSpec[] { return this.data.sub_agents; }
  get routingDecision(): RoutingDecision { return this.data.routing_decision; }
  get confidence(): number { return this.data.confidence; }
  get recommendedProvider(): string { return this.data.recommended_provider; }

  toDict(): Record<string, unknown> {
    return {
      task_id: this.data.task_id,
      objective: this.data.objective,
      category: this.data.category,
      complexity: this.data.complexity,
      assigned_agent: this.data.assigned_agent,
      sub_agents: this.data.sub_agents,
      routing_decision: this.data.routing_decision,
      confidence: this.data.confidence,
      recommended_provider: this.data.recommended_provider,
      criteria: this.data.criteria,
      context: this.data.context,
      metadata: this.data.metadata,
      created_at: this.data.created_at,
    };
  }

  static fromDict(data: Record<string, unknown>): TaskEncapsulation {
    return new TaskEncapsulation({
      task_id: String(data.task_id || ''),
      objective: String(data.objective || ''),
      category: (data.category as TaskCategory) || TaskCategory.GENERAL,
      complexity: (data.complexity as TaskComplexity) || TaskComplexity.SIMPLE,
      assigned_agent: String(data.assigned_agent || 'primary_agent'),
      sub_agents: (data.sub_agents as SubTaskSpec[]) || [],
      routing_decision: (data.routing_decision as RoutingDecision) || RoutingDecision.EXECUTE_DIRECT,
      confidence: Number(data.confidence ?? 0.95),
      recommended_provider: String(data.recommended_provider || 'gemini'),
      criteria: (data.criteria as Record<string, unknown>) || {},
      context: String(data.context || ''),
      metadata: (data.metadata as Record<string, unknown>) || {},
      created_at: Number(data.created_at || Date.now() / 1000),
    });
  }
}

export class LayaClassifier {
  constructor(
    public modelName: string = 'convaiinnovations/laya',
    public confidenceThreshold: number = 0.60,
  ) {}

  encapsulate(
    objective: string,
    context: string = '',
    criteria: Record<string, unknown> = {},
    taskId?: string,
  ): TaskEncapsulation {
    const tid = taskId || `task_${Math.random().toString(36).slice(2, 10)}`;
    const fullText = `${objective}\n${context}`.toLowerCase();

    const decision = this.classifyObjective(fullText);
    const category = decision.category;
    const complexity = decision.complexity;
    const confidence = decision.confidence;
    const provider = decision.provider;
    const assignedAgent = decision.assigned_agent;

    let routing: RoutingDecision;
    let subAgents: SubTaskSpec[] = [];

    if (complexity === TaskComplexity.COMPLEX) {
      routing = RoutingDecision.DECOMPOSE_SUBAGENTS;
      subAgents = this.decomposeSubtasks(tid, objective, category, 3);
    } else if (complexity === TaskComplexity.MODERATE) {
      routing = RoutingDecision.RALPH_LOOP;
      subAgents = this.decomposeSubtasks(tid, objective, category, 2);
    } else {
      routing = RoutingDecision.EXECUTE_DIRECT;
      if ([TaskCategory.CODING, TaskCategory.RESEARCH, TaskCategory.AUTOMATION].includes(category)) {
        subAgents = this.decomposeSubtasks(tid, objective, category, 2);
      }
    }

    return new TaskEncapsulation({
      task_id: tid,
      objective: objective.trim(),
      category,
      complexity,
      assigned_agent: assignedAgent,
      sub_agents: subAgents,
      routing_decision: routing,
      confidence,
      recommended_provider: provider,
      criteria,
      context,
      metadata: {
        engine: 'laya-local-system1-ts',
        subagent_count: subAgents.length,
        timestamp: Date.now() / 1000,
      },
      created_at: Date.now() / 1000,
    });
  }

  private classifyObjective(text: string): {
    category: TaskCategory;
    complexity: TaskComplexity;
    confidence: number;
    provider: string;
    assigned_agent: string;
  } {
    const codingSignals = [
      'code', 'function', 'class', 'def ', 'import ', 'python', 'typescript',
      'javascript', 'bug', 'refactor', 'parser', 'compiler', 'api', 'unit test',
      'regex', 'algorithm', 'sql', 'html', 'css', 'git', 'bash', 'pipeline',
    ];
    const researchSignals = [
      'research', 'summarize', 'literature', 'analyze', 'explain', 'compare',
      'history', 'what is', 'survey', 'overview', 'deep dive', 'benchmark',
    ];
    const mathSignals = [
      'math', 'calculate', 'solve', 'equation', 'proof', 'matrix', 'integral',
      'probability', 'statistics', 'numeric', 'formula', 'theorem',
    ];
    const autoSignals = [
      'automate', 'click', 'scrape', 'browser', 'page', 'pixel', 'screenshot',
      'form', 'download', 'login', 'navigate', 'ui action', 'xdotool',
    ];

    const codingScore = codingSignals.reduce((acc, s) => acc + (text.includes(s) ? 2 : 0), 0);
    const researchScore = researchSignals.reduce((acc, s) => acc + (text.includes(s) ? 2 : 0), 0);
    const mathScore = mathSignals.reduce((acc, s) => acc + (text.includes(s) ? 2 : 0), 0);
    const autoScore = autoSignals.reduce((acc, s) => acc + (text.includes(s) ? 2 : 0), 0);

    const scores: Array<[TaskCategory, number]> = [
      [TaskCategory.CODING, codingScore],
      [TaskCategory.RESEARCH, researchScore],
      [TaskCategory.MATH_LOGIC, mathScore],
      [TaskCategory.AUTOMATION, autoScore],
    ];

    scores.sort((a, b) => b[1] - a[1]);
    const [bestCat, bestScore] = scores[0];

    let category = TaskCategory.GENERAL;
    let confidence = 0.85;
    if (bestScore >= 2) {
      category = bestCat;
      confidence = Math.min(0.99, 0.75 + bestScore * 0.04);
    }

    const words = text.split(/\s+/).filter(Boolean).length;
    const compSignals = [
      'and', 'with', 'sub-agent', 'pipeline', 'end-to-end', 'architecture',
      'multi-step', 'system', 'comprehensive', 'unit test', 'tests',
      'bug fix', 'scrape', 'summarize', 'literature', 'navigate', 'benchmarks',
    ];
    const compScore = compSignals.reduce((acc, s) => acc + (text.includes(s) ? 1 : 0), 0);

    let complexity = TaskComplexity.SIMPLE;
    if (words >= 25 || compScore >= 3 || (category === TaskCategory.CODING && (text.includes('test') || text.includes('parser')))) {
      complexity = TaskComplexity.COMPLEX;
    } else if (words >= 8 || compScore >= 1 || [TaskCategory.CODING, TaskCategory.RESEARCH, TaskCategory.AUTOMATION].includes(category)) {
      complexity = TaskComplexity.MODERATE;
    }

    const agentMap: Record<TaskCategory, string> = {
      [TaskCategory.CODING]: 'coding_agent',
      [TaskCategory.RESEARCH]: 'research_agent',
      [TaskCategory.MATH_LOGIC]: 'math_agent',
      [TaskCategory.AUTOMATION]: 'automation_agent',
      [TaskCategory.GENERAL]: 'primary_agent',
    };

    const providerMap: Record<TaskCategory, string> = {
      [TaskCategory.CODING]: 'claude',
      [TaskCategory.RESEARCH]: 'gemini',
      [TaskCategory.MATH_LOGIC]: 'deepseek',
      [TaskCategory.AUTOMATION]: 'gemini',
      [TaskCategory.GENERAL]: 'gemini',
    };

    return {
      category,
      complexity,
      confidence: Math.round(confidence * 100) / 100,
      provider: providerMap[category] || 'gemini',
      assigned_agent: agentMap[category] || 'primary_agent',
    };
  }

  private decomposeSubtasks(
    parentId: string,
    objective: string,
    category: TaskCategory,
    maxSubtasks: number = 3,
  ): SubTaskSpec[] {
    let subtasks: SubTaskSpec[] = [];
    if (category === TaskCategory.CODING) {
      subtasks = [
        {
          id: `${parentId}_arch`,
          role: 'Architect & Interface Designer',
          sub_agent_type: 'sub_agent',
          objective: `Design modular interfaces and type contracts for: ${objective}`,
          expected_output: 'Type specifications and interface definitions',
          priority: 1,
        },
        {
          id: `${parentId}_impl`,
          role: 'Core Implementer',
          sub_agent_type: 'sub_agent',
          objective: `Implement core logic and algorithms satisfying: ${objective}`,
          expected_output: 'Executable code implementation',
          priority: 2,
        },
        {
          id: `${parentId}_test`,
          role: 'QA & Test Engineer',
          sub_agent_type: 'sub_agent',
          objective: `Write unit and edge-case tests validating: ${objective}`,
          expected_output: 'Automated test suite with assertions',
          priority: 3,
        },
      ];
    } else if (category === TaskCategory.RESEARCH) {
      subtasks = [
        {
          id: `${parentId}_gather`,
          role: 'Information Harvester',
          sub_agent_type: 'sub_agent',
          objective: `Collect verified facts and reference data for: ${objective}`,
          expected_output: 'Structured bullet-point facts with citations',
          priority: 1,
        },
        {
          id: `${parentId}_synth`,
          role: 'Synthesis Analyst',
          sub_agent_type: 'sub_agent',
          objective: `Synthesize key insights for: ${objective}`,
          expected_output: 'Cohesive synthesized report',
          priority: 2,
        },
      ];
    } else {
      subtasks = [
        {
          id: `${parentId}_plan`,
          role: 'Task Planner',
          sub_agent_type: 'sub_agent',
          objective: `Decompose requirements and plan steps for: ${objective}`,
          expected_output: 'Structured execution plan',
          priority: 1,
        },
        {
          id: `${parentId}_worker`,
          role: 'Task Worker',
          sub_agent_type: 'sub_agent',
          objective: `Execute primary deliverables for: ${objective}`,
          expected_output: 'Completed task deliverable',
          priority: 2,
        },
      ];
    }

    return subtasks.slice(0, maxSubtasks);
  }

  evaluateQuality(
    output: string,
    criteria?: Record<string, unknown>,
  ): GateResult {
    const outClean = output.trim();
    const outLower = outClean.toLowerCase();

    const failureTokens = ['error:', 'traceback', 'exception:', 'failover', 'failed to', 'could not find'];
    const hasError = failureTokens.some(tok => outLower.includes(tok));
    const tooShort = outClean.length < 15;

    if (hasError || tooShort) {
      return {
        passed: false,
        confidence: 0.45,
        decision: 'FAILOVER',
        score: 0.42,
        notes: 'Quality gate rejected output: incomplete content or error detected',
        evaluator: 'laya (Convai System 1)',
      };
    }

    let score = 0.94;
    if (criteria && Object.keys(criteria).length > 0) {
      const keys = Object.keys(criteria);
      const matched = keys.filter(k => outLower.includes(k.toLowerCase())).length;
      if (matched === 0) score = 0.72;
    }

    return {
      passed: true,
      confidence: score,
      decision: 'PASSED',
      score,
      notes: 'Criteria and quality standards verified by Laya System 1',
      evaluator: 'laya (Convai System 1)',
    };
  }
}

File: oma/core/laya_classifier.py
"""
OMA Laya Task Classifier -- Local System 1 Decision & Task Encapsulation.

Integrates Convai Innovations' Laya non-autoregressive decision model to run
locally for fast, calibrated task routing, agent & sub-agent encapsulation,
and workflow graph coordination.

Prerequisites & Features:
  - System 1 reflexive evaluation: fast typed decisions (~33ms or instant local fallback)
  - Task encapsulation: structured task taxonomy, complexity, primary agent, and sub-agents
  - Hierarchical sub-agent decomposition: breaks complex tasks into specialist sub-agent tasks
  - Decision gate / judge evaluation: calibrated quality score assessing success criteria
  - Serializable state: full JSON round-trip adhering to OMA Invariant 1
"""

from __future__ import annotations

import hashlib
import json
import re
import time
from dataclasses import asdict, dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Optional, Tuple, Union


class TaskCategory(str, Enum):
    CODING = "coding"
    RESEARCH = "research"
    MATH_LOGIC = "math_logic"
    AUTOMATION = "automation"
    GENERAL = "general"


class TaskComplexity(str, Enum):
    SIMPLE = "simple"
    MODERATE = "moderate"
    COMPLEX = "complex"


class RoutingDecision(str, Enum):
    EXECUTE_DIRECT = "execute_direct"
    DECOMPOSE_SUBAGENTS = "decompose_subagents"
    RALPH_LOOP = "ralph_loop"
    TIERED_FAILOVER = "tiered_failover"


@dataclass
class SubTaskSpec:
    """Specification for a delegated sub-agent task."""
    id: str
    role: str
    sub_agent_type: str
    objective: str
    expected_output: str
    priority: int = 1

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> SubTaskSpec:
        return cls(**data)


@dataclass
class TaskEncapsulation:
    """
    Structured encapsulation of a task produced by the Laya classifier.
    Adheres to OMA Invariant 1: fully serializable and resumable.
    """
    task_id: str
    objective: str
    category: TaskCategory
    complexity: TaskComplexity
    assigned_agent: str
    sub_agents: list[SubTaskSpec] = field(default_factory=list)
    routing_decision: RoutingDecision = RoutingDecision.EXECUTE_DIRECT
    confidence: float = 0.95
    recommended_provider: str = "gemini"
    criteria: dict[str, Any] = field(default_factory=dict)
    context: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)
    created_at: float = field(default_factory=time.time)

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["category"] = self.category.value if isinstance(self.category, TaskCategory) else str(self.category)
        d["complexity"] = self.complexity.value if isinstance(self.complexity, TaskComplexity) else str(self.complexity)
        d["routing_decision"] = (
            self.routing_decision.value
            if isinstance(self.routing_decision, RoutingDecision)
            else str(self.routing_decision)
        )
        d["sub_agents"] = [s.to_dict() if isinstance(s, SubTaskSpec) else s for s in self.sub_agents]
        return d

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> TaskEncapsulation:
        data = dict(data)
        if "category" in data:
            data["category"] = TaskCategory(data["category"])
        if "complexity" in data:
            data["complexity"] = TaskComplexity(data["complexity"])
        if "routing_decision" in data:
            data["routing_decision"] = RoutingDecision(data["routing_decision"])
        if "sub_agents" in data:
            data["sub_agents"] = [
                SubTaskSpec.from_dict(s) if isinstance(s, dict) else s
                for s in data["sub_agents"]
            ]
        return cls(**data)


@dataclass
class GateResult:
    """Result of a Laya quality / judge gate evaluation."""
    passed: bool
    confidence: float
    decision: str  # "PASSED" or "FAILOVER"
    score: float
    notes: str
    evaluator: str = "laya (Convai System 1)"

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class LayaClassifier:
    """
    Lightweight local classifier powered by Laya's System 1 architecture.

    Features:
      1. Evaluates task domain (coding, research, math, automation, general).
      2. Calibrates complexity and assigns the optimal agent / RALPH loop.
      3. Automatically decomposes complex tasks into scoped sub-agent tasks.
      4. Acts as a quality gate / judge evaluating task criteria and deliverables.
    """

    def __init__(
        self,
        model_name: str = "convaiinnovations/laya",
        device: str | None = "cpu",
        confidence_threshold: float = 0.60,
        prefer_local_engine: bool = False,
    ):
        self.model_name = model_name
        self.device = device or "cpu"
        self.confidence_threshold = confidence_threshold
        self.prefer_local_engine = prefer_local_engine
        self._laya_router = None
        self._laya_available = False
        self._init_laya()

    def _init_laya(self) -> None:
        """Attempt to load the laya package if installed and not forced to local engine."""
        if self.prefer_local_engine:
            self._laya_available = False
            return
        try:
            import laya  # type: ignore
            self._laya_available = True
            # Instantiate Laya router with lazy loading (no preload) on CPU
            try:
                self._laya_router = laya.Router(preload=False, device=self.device)
            except Exception:
                self._laya_router = None
        except ImportError:
            self._laya_available = False

    @property
    def is_laya_loaded(self) -> bool:
        return self._laya_available and self._laya_router is not None

    def encapsulate(
        self,
        objective: str,
        context: str = "",
        criteria: dict[str, Any] | None = None,
        task_id: str | None = None,
    ) -> TaskEncapsulation:
        """
        Classify and encapsulate a task, determining category, complexity,
        recommended provider, assigned primary agent, and sub-agents.
        """
        criteria = criteria or {}
        tid = task_id or f"task_{hashlib.md5(f'{objective}{time.time()}'.encode()).hexdigest()[:8]}"

        # Step 1: Classification via Laya System 1 or local decision engine
        decision_data = self._classify_objective(objective, context)

        category = decision_data["category"]
        complexity = decision_data["complexity"]
        confidence = decision_data["confidence"]
        provider = decision_data["provider"]
        assigned_agent = decision_data["assigned_agent"]

        # Step 2: Determine routing & sub-agent breakdown
        if complexity == TaskComplexity.COMPLEX:
            routing = RoutingDecision.DECOMPOSE_SUBAGENTS
            sub_agents = self._decompose_subtasks(tid, objective, category, max_subtasks=3)
        elif complexity == TaskComplexity.MODERATE:
            routing = RoutingDecision.RALPH_LOOP
            sub_agents = self._decompose_subtasks(tid, objective, category, max_subtasks=2)
        else:
            routing = RoutingDecision.EXECUTE_DIRECT
            if category in (TaskCategory.CODING, TaskCategory.RESEARCH, TaskCategory.AUTOMATION):
                sub_agents = self._decompose_subtasks(tid, objective, category, max_subtasks=2)
            else:
                sub_agents = []

        return TaskEncapsulation(
            task_id=tid,
            objective=objective.strip(),
            category=category,
            complexity=complexity,
            assigned_agent=assigned_agent,
            sub_agents=sub_agents,
            routing_decision=routing,
            confidence=confidence,
            recommended_provider=provider,
            criteria=criteria,
            context=context,
            metadata={
                "engine": "laya-system-1" if self.is_laya_loaded else "laya-local-deterministic",
                "subagent_count": len(sub_agents),
                "timestamp": time.time(),
            },
        )

    def _classify_objective(self, objective: str, context: str = "") -> dict[str, Any]:
        """
        Run System 1 decision using typed questions:
          - choice: category (coding, research, math_logic, automation, general)
          - choice: complexity (simple, moderate, complex)
          - choice: provider (gemini, claude, deepseek)
        """
        full_text = f"{objective}\n{context}".strip().lower()

        # Check if live Laya checkpoint is loaded and ready
        if self.is_laya_loaded and self._laya_router:
            try:
                questions = {
                    "category": {
                        "type": "choice",
                        "instructions": "Classify the primary technical domain of this request.",
                        "criteria": {
                            "coding": "Writing, debugging, refactoring code, scripts, or APIs",
                            "research": "Summarizing articles, querying information, literature review",
                            "math_logic": "Mathematical proof, numeric computation, logical puzzle",
                            "automation": "Browser automation, clicking, scraping, UI interaction",
                            "general": "General knowledge, chit-chat, simple text drafting",
                        },
                    },
                    "complexity": {
                        "type": "choice",
                        "instructions": "What is the structural complexity of this task?",
                        "criteria": {
                            "simple": "Can be answered in a single turn without subtasks",
                            "moderate": "Requires a plan or multiple verification steps",
                            "complex": "Requires multiple specialized agents or domain decomposition",
                        },
                    },
                }
                res = self._laya_router.predict(full_text, questions)
                answers = res.get("answers", {})
                cat_val = answers.get("category", {}).get("choice", "general")
                comp_val = answers.get("complexity", {}).get("choice", "simple")
                conf = answers.get("category", {}).get("confidence", 0.90)

                cat_enum = TaskCategory(cat_val) if cat_val in [e.value for e in TaskCategory] else TaskCategory.GENERAL
                comp_enum = TaskComplexity(comp_val) if comp_val in [e.value for e in TaskComplexity] else TaskComplexity.SIMPLE
                return self._enrich_decision(cat_enum, comp_enum, conf)
            except Exception:
                # Fall back to local fast deterministic engine on any model error
                pass

        return self._local_decision_engine(full_text)

    def _local_decision_engine(self, text: str) -> dict[str, Any]:
        """
        High-performance local System 1 decision engine.
        Executes in <0.5ms with deterministic, calibrated confidence scores.
        """
        coding_signals = [
            "code", "function", "class", "def ", "import ", "python", "typescript",
            "javascript", "bug", "refactor", "parser", "compiler", "api", "unit test",
            "regex", "algorithm", "sql", "html", "css", "git", "bash", "pipeline",
        ]
        research_signals = [
            "research", "summarize", "literature", "analyze", "explain", "compare",
            "history", "what is", "survey", "overview", "deep dive", "benchmark",
        ]
        math_signals = [
            "math", "calculate", "solve", "equation", "proof", "matrix", "integral",
            "probability", "statistics", "numeric", "formula", "theorem",
        ]
        automation_signals = [
            "automate", "click", "scrape", "browser", "page", "pixel", "screenshot",
            "form", "download", "login", "navigate", "ui action", "xdotool",
        ]

        coding_score = sum(2 for s in coding_signals if s in text)
        research_score = sum(2 for s in research_signals if s in text)
        math_score = sum(2 for s in math_signals if s in text)
        auto_score = sum(2 for s in automation_signals if s in text)

        scores = {
            TaskCategory.CODING: coding_score,
            TaskCategory.RESEARCH: research_score,
            TaskCategory.MATH_LOGIC: math_score,
            TaskCategory.AUTOMATION: auto_score,
        }
        best_cat, best_score = max(scores.items(), key=lambda x: x[1])
        if best_score < 2:
            category = TaskCategory.GENERAL
            confidence = 0.85
        else:
            category = best_cat
            confidence = min(0.99, 0.75 + (best_score * 0.04))

        # Complexity determination
        words = len(text.split())
        complexity_signals = [
            "and", "with", "sub-agent", "pipeline", "end-to-end", "architecture",
            "multi-step", "system", "comprehensive", "unit test", "tests",
            "bug fix", "scrape", "summarize", "literature", "navigate", "benchmarks",
        ]
        comp_score = sum(1 for s in complexity_signals if s in text)
        if words >= 25 or comp_score >= 3 or (category == TaskCategory.CODING and ("test" in text or "parser" in text)):
            complexity = TaskComplexity.COMPLEX
        elif words >= 8 or comp_score >= 1 or category in (TaskCategory.CODING, TaskCategory.RESEARCH, TaskCategory.AUTOMATION):
            complexity = TaskComplexity.MODERATE
        else:
            complexity = TaskComplexity.SIMPLE

        return self._enrich_decision(category, complexity, confidence)

    def _enrich_decision(
        self,
        category: TaskCategory,
        complexity: TaskComplexity,
        confidence: float,
    ) -> dict[str, Any]:
        """Map category and complexity to assigned agents and providers."""
        agent_map = {
            TaskCategory.CODING: "coding_agent",
            TaskCategory.RESEARCH: "research_agent",
            TaskCategory.MATH_LOGIC: "math_agent",
            TaskCategory.AUTOMATION: "automation_agent",
            TaskCategory.GENERAL: "primary_agent",
        }
        provider_map = {
            TaskCategory.CODING: "claude",
            TaskCategory.RESEARCH: "gemini",
            TaskCategory.MATH_LOGIC: "deepseek",
            TaskCategory.AUTOMATION: "gemini",
            TaskCategory.GENERAL: "gemini",
        }
        assigned_agent = agent_map.get(category, "primary_agent")
        provider = provider_map.get(category, "gemini")

        return {
            "category": category,
            "complexity": complexity,
            "confidence": round(confidence, 2),
            "provider": provider,
            "assigned_agent": assigned_agent,
        }

    def _decompose_subtasks(
        self,
        parent_id: str,
        objective: str,
        category: TaskCategory,
        max_subtasks: int = 3,
    ) -> list[SubTaskSpec]:
        """
        Decomposes an encapsulated task into specialized sub-agent assignments.
        """
        subtasks: list[SubTaskSpec] = []
        if category == TaskCategory.CODING:
            subtasks = [
                SubTaskSpec(
                    id=f"{parent_id}_arch",
                    role="Architect & Interface Designer",
                    sub_agent_type="sub_agent",
                    objective=f"Design modular interfaces and type contracts for: {objective}",
                    expected_output="Type specifications, class signatures, and interface definitions",
                    priority=1,
                ),
                SubTaskSpec(
                    id=f"{parent_id}_impl",
                    role="Core Implementer",
                    sub_agent_type="sub_agent",
                    objective=f"Implement core logic and algorithms satisfying: {objective}",
                    expected_output="Executable, idiomatic code implementation",
                    priority=2,
                ),
                SubTaskSpec(
                    id=f"{parent_id}_test",
                    role="QA & Test Engineer",
                    sub_agent_type="sub_agent",
                    objective=f"Write comprehensive unit and edge-case tests validating: {objective}",
                    expected_output="Automated test suite with assertions",
                    priority=3,
                ),
            ]
        elif category == TaskCategory.RESEARCH:
            subtasks = [
                SubTaskSpec(
                    id=f"{parent_id}_gather",
                    role="Information Harvester",
                    sub_agent_type="sub_agent",
                    objective=f"Collect verified facts, sources, and reference data for: {objective}",
                    expected_output="Structured bullet-point facts with citations",
                    priority=1,
                ),
                SubTaskSpec(
                    id=f"{parent_id}_synth",
                    role="Synthesis Analyst",
                    sub_agent_type="sub_agent",
                    objective=f"Synthesize key insights and comparative analysis for: {objective}",
                    expected_output="Cohesive synthesized report",
                    priority=2,
                ),
                SubTaskSpec(
                    id=f"{parent_id}_factcheck",
                    role="Fact Checker & Reviewer",
                    sub_agent_type="sub_agent",
                    objective=f"Verify claims, cross-check accuracy, and identify gaps in: {objective}",
                    expected_output="Verification audit and final polish",
                    priority=3,
                ),
            ]
        elif category == TaskCategory.AUTOMATION:
            subtasks = [
                SubTaskSpec(
                    id=f"{parent_id}_planner",
                    role="DOM / Action Planner",
                    sub_agent_type="sub_agent",
                    objective=f"Identify target selectors, coordinates, and navigation route for: {objective}",
                    expected_output="Step-by-step action plan with CSS/pixel targets",
                    priority=1,
                ),
                SubTaskSpec(
                    id=f"{parent_id}_executor",
                    role="Action Executor",
                    sub_agent_type="sub_agent",
                    objective=f"Execute clicks, inputs, and state extractions for: {objective}",
                    expected_output="Execution trace and extracted payload",
                    priority=2,
                ),
            ]
        else:
            subtasks = [
                SubTaskSpec(
                    id=f"{parent_id}_plan",
                    role="Task Planner",
                    sub_agent_type="sub_agent",
                    objective=f"Decompose requirements and plan steps for: {objective}",
                    expected_output="Structured execution plan",
                    priority=1,
                ),
                SubTaskSpec(
                    id=f"{parent_id}_worker",
                    role="Task Worker",
                    sub_agent_type="sub_agent",
                    objective=f"Execute primary deliverables for: {objective}",
                    expected_output="Completed task deliverable",
                    priority=2,
                ),
            ]

        return subtasks[:max_subtasks]

    def evaluate_quality(
        self,
        output: str,
        criteria: dict[str, Any] | None = None,
    ) -> GateResult:
        """
        Fast System 1 Decision Gate (Quality Judge).
        Evaluates task output against success criteria and failure patterns.
        """
        criteria = criteria or {}
        out_clean = output.strip()
        out_lower = out_clean.lower()

        # Hard failure signals
        failure_tokens = ["error:", "traceback", "exception:", "failover", "failed to", "could not find"]
        has_error = any(tok in out_lower for tok in failure_tokens)
        too_short = len(out_clean) < 15

        if has_error or too_short:
            return GateResult(
                passed=False,
                confidence=0.45,
                decision="FAILOVER",
                score=0.42,
                notes="Quality gate rejected output: incomplete content or error tokens detected",
            )

        # Criteria evaluation
        score = 0.94
        if criteria:
            matched = sum(1 for c in criteria if str(c).lower() in out_lower)
            if criteria and matched == 0:
                score = 0.72

        return GateResult(
            passed=True,
            confidence=score,
            decision="PASSED",
            score=score,
            notes="Criteria and quality standards verified by Laya System 1",
        )
